pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java

```diff
@@ -0,0 +1,225 @@
+                    String errorMsg = String.format(
+                            "Failed to look up a broker registry:%s for bundle:%s", broker, bundle);
+                    log.error(errorMsg);
+                    throw new IllegalStateException(errorMsg);
+                }
+                return CompletableFuture.completedFuture(brokerLookupData);
+            }));
+        });
+        future.whenComplete((r, t) -> lookupRequests.remove(bundle));
+        return future;
+    }
+
+    public CompletableFuture<Optional<String>> selectAsync(ServiceUnitId bundle) {
+        BrokerRegistry brokerRegistry = getBrokerRegistry();
+        return brokerRegistry.getAvailableBrokerLookupDataAsync()
+                .thenCompose(availableBrokerCandidates -> {
+                    // TODO: Support isolation policies
+                    LoadManagerContext context = this.getContext();
+
+                    // Filter out brokers that do not meet the rules.
+                    List<BrokerFilter> filterPipeline = getBrokerFilterPipeline();
+                    for (final BrokerFilter filter : filterPipeline) {
+                        try {
+                            filter.filter(availableBrokerCandidates, context);
+                        } catch (BrokerFilterException e) {
+                            log.error("Failed to filter out brokers.", e);
+                            return CompletableFuture.completedFuture(Optional.empty());
+                        }
+                    }
+                    if (availableBrokerCandidates.isEmpty()) {
+                        return CompletableFuture.completedFuture(Optional.empty());
+                    }
+                    ArrayList<String> candidateBrokers = new ArrayList<>(availableBrokerCandidates.keySet());
+
+                    return CompletableFuture.completedFuture(
+                            getBrokerSelectionStrategy().select(candidateBrokers, bundle, context));
+                });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> checkOwnershipAsync(Optional<ServiceUnitId> topic, ServiceUnitId bundleUnit) {
+        final String bundle = bundleUnit.toString();
+        CompletableFuture<Optional<String>> owner;
+        if (topic.isPresent() && isInternalTopic(topic.get().toString())) {
+            owner = serviceUnitStateChannel.getChannelOwnerAsync();
+        } else {
+            owner = serviceUnitStateChannel.getOwnerAsync(bundle);
+        }
+
+        return owner.thenApply(broker -> brokerRegistry.getBrokerId().equals(broker.orElse(null)));
+    }
+
+    @Override
+    public synchronized void close() throws PulsarServerException {
+        if (!this.started) {
+            return;
+        }
+        try {
+            this.brokerRegistry.close();
+        } catch (Exception e) {
+            throw new PulsarServerException(e);
+        }
+        this.serviceUnitStateChannel.close();
+        this.started = false;
```

Comment:
Move it to `finally` because when `serviceUnitStateChannel.close()` throws an `PulsarServerException`, it will be skipped.

Reply:
Additional, `serviceUnitStateChannel.close()` could be skipped when `brokerRegistry.close()` throws an exception (BTW, why did you catch `Exception`?)

If you don't care which exception will be thrown, you can simply write like:

```java
        try {
            this.brokerRegistry.close();
        } finally {
            try {
                this.serviceUnitStateChannel.close();
            } finally {
                this.started = false;
            }
        }
```

Otherwise, it would be a little complicated to combine two `PulsarServerException` into a single `PulsarServerException`.